gnmidiff/setrequest.go

```diff
@@ -0,0 +1,236 @@
+		}
+	}
+	for pathB, vB := range intentB.Updates {
+		if _, ok := intentA.Updates[pathB]; !ok {
+			diff.BOnlyUpdates[pathB] = vB
+		}
+	}
+	return diff, nil
+}
+
+type setRequestIntent struct {
+	Deletes map[string]struct{}
+	Updates map[string]interface{}
+}
+
+// minimalSetRequestIntent returns a unique and minimal intent for a SetRequest.
+//
+// Currently, support is only for SetRequests without any delete paths, and
+// replace and updates that don't have conflicting leaf values. If not
+// supported, then an error will be returned.
+func minimalSetRequestIntent(req *gpb.SetRequest) (setRequestIntent, error) {
+	if req == nil {
+		req = &gpb.SetRequest{}
+	}
+	if len(req.Delete) > 0 {
+		return setRequestIntent{}, fmt.Errorf("gnmidiff: delete paths are not supported.")
+	}
+	intent := setRequestIntent{
+		Deletes: map[string]struct{}{},
+		Updates: map[string]interface{}{},
+	}
+	// NOTE: This simple trie will not work if we intend to check conflicts with wildcard deletion paths.
+	t := trie.New()
+	for _, upd := range req.Replace {
+		path, err := ygot.PathToString(upd.Path)
+		if err != nil {
+			return setRequestIntent{}, fmt.Errorf("gnmidiff: %v", err)
+		}
+		if _, ok := intent.Deletes[path]; ok {
+			return setRequestIntent{}, fmt.Errorf("gnmidiff: conflicting replaces in SetRequest: %v", path)
+		}
+		intent.Deletes[path] = struct{}{}
+		t.Add(path, nil)
+
+		if err := populateUpdates(&intent, path, upd.GetVal()); err != nil {
+			return setRequestIntent{}, err
+		}
+	}
+
+	// Do prefix match to check for conflicting replace paths.
+	for _, path := range t.Keys() {
+		if matches := t.PrefixSearch(path); len(matches) >= 2 {
+			return setRequestIntent{}, fmt.Errorf("gnmidiff: conflicting replaces in SetRequest: %v", matches)
+		}
+	}
+
+	for _, upd := range req.Update {
+		path, err := ygot.PathToString(upd.Path)
+		if err != nil {
+			return setRequestIntent{}, fmt.Errorf("gnmidiff: %v", err)
+		}
+
+		if err := populateUpdates(&intent, path, upd.GetVal()); err != nil {
+			return setRequestIntent{}, err
+		}
+	}
+
+	t = trie.New()
+	for path := range intent.Updates {
+		t.Add(path, nil)
+	}
+
+	// Do prefix match to check for conflicting update paths.
+	for _, path := range t.Keys() {
+		if matches := t.PrefixSearch(path); len(matches) >= 2 {
+			return setRequestIntent{}, fmt.Errorf("gnmidiff: bad SetRequest, there are leaf updates that have a prefix match: %v", matches)
+		}
+	}
+
+	return intent, nil
+}
+
+// populateUpdates populates all leaf updates at the given path into the intent.
+//
+// Note: The input path must NOT end with "/".
+//
+// e.g. /a for b/c="foo" would introduce an update of /a/b/c="foo" into the intent.
+func populateUpdates(intent *setRequestIntent, path string, tv *gpb.TypedValue) error {
+	if len(path) > 0 && path[len(path)-1] == '/' {
+		return fmt.Errorf("gnmidiff: invalid input path %q, must not end with \"/\"", path)
+	}
+	var leafVal interface{}
+	isLeaf := true
+	switch tv.GetValue().(type) {
```

Comment:
You could use gnmi's `value.ToScalar` if you handle the non-scalar values first I think.

Reply:
`ToScalar` maps to a Go native type whereas a JSON type helps for consistency in comparison. e.g. for a JSON `5` is that a double or an integer?

Reply:
I've split this code to https://github.com/openconfig/ygot/pull/775 for easier review.